### PR TITLE
ref: Add CacheEntry type and use it for SymCaches and PpdbCaches

### PR DESCRIPTION
**Pull request by loewenheim**
This adds a new enum `CacheEntry` that is intended to eventually replace the current representation of cache entries as (essentially) a pair of `(CacheStatus, ByteView)` and the myriad different errors. As a first step, we use it in `SymCacheActor::fetch` and `PortablePdbCacheActor::fetch` and consequently also in the `module_lookup` module. The code in the latter becomes slightly simpler, in my opinion, but the big win is that we parse the caches only once and from then on pass around the parsed object instead of using a `ByteView` that has to be parsed over and over again.

This currently requires a development version of symbolic because some additional traits have to be implemented for `PortablePdbCache`.

EDIT: The most recent version replaces the custom `CacheEntry` enum with a `Result`. The semantics stay exactly the same.

### CODE REVIEW

File: crates/symbolicator-service/src/cache.rs [hunk 1]
```diff
@@ -119,6 +120,174 @@ impl CacheStatus {
...
+    /// An unexpected error in symbolicator itself.
+    InternalError,
+}
+
+impl<T> CacheEntry<T> {
+    const PERMISSION_DENIED_MARKER: &[u8] = b"permissiondenied";
+    const TIMEOUT_MARKER: &[u8] = b"timeout";
+    const DOWNLOAD_ERROR_MARKER: &[u8] = b"downloaderror";
+    const MALFORMED_MARKER: &[u8] = b"malformed";
+    const INTERNAL_ERROR_MARKER: &[u8] = b"internalerror";
```
Comment (Swatinem) [inline crates/symbolicator-service/src/cache.rs:157]: I don’t think we will ever persist these, as this is meant for errors that happen during persisting or reading from disk.

File: crates/symbolicator-service/src/cache.rs [hunk 2]
```diff
@@ -119,6 +120,174 @@ impl CacheStatus {
...
+}
+
+impl From<(CacheStatus, ByteView<'static>)> for CacheEntry<ByteView<'static>> {
+    fn from((status, content): (CacheStatus, ByteView<'static>)) -> Self {
+        match status {
+            CacheStatus::Positive => Self::AllGood(content),
+            CacheStatus::Negative => Self::NotFound,
+            CacheStatus::Malformed(s) => Self::Malformed(s),
+            CacheStatus::CacheSpecificError(message) => {
+                if let Some(details) = message.strip_prefix("missing permissions for file") {
```
Comment (Swatinem) [inline crates/symbolicator-service/src/cache.rs:279]: can we use whatever constant is used inside the `DownloadError`, or is it not that simple?

File: crates/symbolicator-service/src/services/ppdb_caches.rs [hunk 3]
```diff
@@ -68,6 +76,21 @@ pub enum PortablePdbCacheError {
...
     Timeout,
 }
+
+impl From<&PortablePdbCacheError> for CacheEntry<OwnedPortablePdbCache> {
+    fn from(error: &PortablePdbCacheError) -> Self {
+        match *error {
+            PortablePdbCacheError::Io(_)
+            | PortablePdbCacheError::Parsing(_)
+            | PortablePdbCacheError::PortablePdbParsing(_)
+            | PortablePdbCacheError::Writing(_) => Self::InternalError,
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/ppdb_caches.rs:86]: should we log these at least?

File: crates/symbolicator-service/src/services/ppdb_caches.rs [hunk 4]
```diff
@@ -142,26 +169,38 @@ impl PortablePdbCacheActor {
...
             Some(handle) => {
-                self.ppdb_caches
+                match self
+                    .ppdb_caches
                     .compute_memoized(FetchPortablePdbCacheInternal {
                         request,
                         objects_actor: self.objects.clone(),
                         object_meta: handle,
-                        candidates,
+                        candidates: candidates.clone(),
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/ppdb_caches.rs:184]: Improvement opportunity: We don’t need to pass these along through the `FetchPortablePdbCacheInternal` / `PortablePdbCacheFile` anymore now, as we can just use the `candidates` that we already have.

This is also more correct. As `compute_memoized` does request coalescing, there is a slim chance that we end up with `candidates` from a different request.

File: crates/symbolicator-service/src/services/ppdb_caches.rs [hunk 5]
```diff
@@ -142,26 +169,38 @@ impl PortablePdbCacheActor {
...
+                            data,
+                            candidates,
+                            features,
+                            ..
+                        } = Arc::try_unwrap(ppdb_cache_file).unwrap_or_else(|arc| (*arc).clone());
+                        let cache_entry = CacheEntry::from((status, data));
+                        let mapped = cache_entry.try_map(parse_ppdb_cache_owned);
+                        (mapped, candidates, features)
+                    }
+                    Err(e) => (e.as_ref().into(), candidates, ObjectFeatures::default()),
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/ppdb_caches.rs:200]: in theory the object meta handle should have `features`. Similar thing applies here as above: We don’t need to pass those along through the whole `FetchPortablePdbCacheInternal`.

File: crates/symbolicator-service/src/services/symcaches/mod.rs [hunk 6]
```diff
@@ -375,23 +402,33 @@ impl SymCacheActor {
...
+                        candidates: candidates.clone(),
                     })
                     .await
+                {
+                    Ok(symcache_file) => {
+                        let SymCacheFile {
+                            status,
+                            data,
+                            candidates,
+                            features,
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/symcaches/mod.rs:421]: I believe the same thing as for ppdb should apply here for symcache as well.

File: crates/symbolicator-service/src/services/symbolication/symbolication.rs [hunk 7]
```diff
@@ -135,31 +136,27 @@ fn symbolicate_frame(
...
+            signal,
+            frame,
+            index,
+        ),
+        CacheEntry::AllGood(CacheFileEntry::PortablePdbCache(ppdb_cache)) => {
+            symbolicate_dotnet_frame(ppdb_cache.get(), frame, index)
         }
-        None => Err(FrameStatus::Missing),
+        CacheEntry::Malformed(_) => Err(FrameStatus::Malformed),
+        _ => Err(FrameStatus::Missing),
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/symbolication/symbolication.rs:151]: Does `FrameStatus` have more options that we can use now?

File: crates/symbolicator-service/src/services/symbolication/module_lookup.rs [hunk 8]
```diff
@@ -9,18 +9,33 @@ use symbolic::common::{ByteView, SelfCell};
...
 use super::object_id_from_object_info;
 
+fn object_file_status_from_cache_entry<T>(cache_entry: &CacheEntry<T>) -> ObjectFileStatus {
+    match cache_entry {
+        CacheEntry::AllGood(_) => ObjectFileStatus::Found,
+        CacheEntry::NotFound => ObjectFileStatus::Missing,
+        CacheEntry::PermissionDenied(_) | CacheEntry::DownloadError(_) => {
+            ObjectFileStatus::FetchingFailed
+        }
+        CacheEntry::Timeout(_) => ObjectFileStatus::Timeout,
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/symbolication/module_lookup.rs:33]: 🎉

**Review by Swatinem — APPROVED** [review]

File: crates/symbolicator-service/src/cache.rs (hunk 2 above)
Comment (loewenheim) [reply crates/symbolicator-service/src/cache.rs:279]: `DownloadError` uses `&Self::Permissions.to_string()`. We could use that, but tying another completely unrelated type into this conversion seems wrong to me.

File: crates/symbolicator-service/src/services/ppdb_caches.rs (hunk 3 above)
Comment (loewenheim) [reply crates/symbolicator-service/src/services/ppdb_caches.rs:86]: Good idea, yeah.

File: crates/symbolicator-service/src/services/symbolication/symbolication.rs (hunk 7 above)
Comment (loewenheim) [reply crates/symbolicator-service/src/services/symbolication/symbolication.rs:151]: The definition of `FrameStatus` is
```
pub enum FrameStatus {
    /// The frame was symbolicated successfully.
    Symbolicated,
    /// The symbol (i.e. function) was not found within the debug file.
    MissingSymbol,
    /// No debug image is specified for the address of the frame.
    UnknownImage,
    /// The debug file could not be retrieved from any of the sources.
    Missing,
    /// The retrieved debug file could not be processed.
    Malformed,
}
```
I don't think any of these except for `Missing` and `Malformed` are useful here.

File: crates/symbolicator-service/src/cache.rs (hunk 2 above)
Comment (Swatinem) [reply crates/symbolicator-service/src/cache.rs:274]: https://github.com/getsentry/symbolicator/blob/7bad96c1908a9772d3d259a9808ab26f1417bf98/crates/symbolicator-service/src/services/download/mod.rs#L104-L105

lol, current code uses `to_string()` just to check for that condition

**Review by Swatinem — APPROVED** [review]

**Review by Swatinem — APPROVED** [review]

**Pull request merged**